### ISSUE TITLE
[C++ Frontend] TORCH_ARG should create public members to allow for named aggregate initialization

### ISSUE THREAD

**Issue by supsm** [pytorch/pytorch#108794]
### 🚀 The feature, motivation and pitch

The current implementation of many options in the high-level C++ frontend is rather clunky, due to the desire to replicate python's named arguments. I will be using `torch::data::DataLoaderOptions` as an example, but the issue applies to almost all functions. Here is an example from [the official tutorial docs](https://pytorch.org/tutorials/advanced/cpp_frontend.html): 
```c++
auto data_loader = torch::data::make_data_loader(
    std::move(dataset),
    torch::data::DataLoaderOptions().batch_size(kBatchSize).workers(2));
```
The use of naming the entire type, `torch::data::DataLoaderOptions`, makes the code unnecessary long and intimidating. It is more apparent when only on one line:
```c++
auto data_loader = torch::data::make_data_loader(std::move(dataset), torch::data::DataLoaderOptions().batch_size(kBatchSize).workers(2));
```
C++20's [designated initializers] would make this easier, as it allows braced initialization of data structs by naming individual specified members, leaving others default initialized. It is significantly cleaner and more concise, while also better aligning with the python API:
```c++
auto data_loader = torch::data::make_data_loader(std::move(dataset), { .batch_size = kBatchSize, .workers = 2 });
```
`DataLoaderOptions` and other options do not allow for aggregate initialization since it relies on the [`TORCH_ARG`](https://github.com/pytorch/pytorch/blob/main/torch/csrc/api/include/torch/arg.h) macro, which creates the args as private members, and since some have a user-defined constructor (e.g. `DataLoaderOptions(size_t batch_size)`). From what I can tell, creating them as private members has no advantage and is only to conform to a traditional OOP factory paradigm, so this change should be trivial. The other change of removing the constructor may be less favorable, but I believe the benefit to significantly outweigh the costs. Breaking backwards compatibility is a big deal, but even on older language standards, the fix for existing codebases is simply changing `torch::data::DataLoaderOptions(batch_size)` to `torch::data::DataLoaderOptions().batch_size(batch_size)`, which also leads to more readable code.

**Comment by r83575** [comment]
Hi, I'd like to work on this and open a PR to address it.

Please let me know if there are any specific constraints or preferences I should be aware of before starting.

Thanks!
